tests/test_request.c

```diff
@@ -194,3 +199,78 @@ TEST_CASE(request_erase_headers) {
     aws_http_request_destroy(request);
     return AWS_OP_SUCCESS;
 }
+
+/* Do every operation that involves allocating some memory */
+int s_request_handles_oom_attempt(struct aws_http_request *request) {
+    ASSERT_NOT_NULL(request);
+
+    /* Set, and then overwrite, method and path */
+    ASSERT_SUCCESS(aws_http_request_set_method(request, aws_byte_cursor_from_c_str("POST")));
+    ASSERT_SUCCESS(aws_http_request_set_path(request, aws_byte_cursor_from_c_str("/")));
+    ASSERT_SUCCESS(aws_http_request_set_method(request, aws_byte_cursor_from_c_str("GET")));
+    ASSERT_SUCCESS(aws_http_request_set_path(request, aws_byte_cursor_from_c_str("/chat")));
+
+    /* Add a lot of headers, enough to force the underlying array-list to expand.
+     * (just loop through the list above again and again) */
+    char name_buf[16];
+    char value_buf[16];
+    for (size_t i = 0; i < 128; ++i) {
+        snprintf(name_buf, sizeof(name_buf), "Name-%zu", i);
+        snprintf(name_buf, sizeof(name_buf), "Value-%zu", i);
+        struct aws_http_header header = {.name = aws_byte_cursor_from_c_str(name_buf),
+                                         .value = aws_byte_cursor_from_c_str(value_buf)};
+        ASSERT_SUCCESS(aws_http_request_add_header(request, header));
+    }
+
+    /* Overwrite all the headers */
+    for (size_t i = 0; i < 128; ++i) {
+        snprintf(name_buf, sizeof(name_buf), "New-Name-%zu", i);
+        snprintf(name_buf, sizeof(name_buf), "New-Value-%zu", i);
+        struct aws_http_header header = {.name = aws_byte_cursor_from_c_str(name_buf),
+                                         .value = aws_byte_cursor_from_c_str(value_buf)};
+        ASSERT_SUCCESS(aws_http_request_set_header(request, header, i));
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+TEST_CASE(request_handles_oom) {
+    (void)ctx;
+    struct aws_allocator timebomb_alloc;
+    ASSERT_SUCCESS(aws_timebomb_allocator_init(&timebomb_alloc, allocator, SIZE_MAX));
+
+    bool test_succeeded = false;
+    size_t allocations_until_failure;
+    for (allocations_until_failure = 0; allocations_until_failure < 10000; ++allocations_until_failure) {
```

Comment:
there's only like 300-ish possible allocations, is 10K necessary?

Reply:
DUDE IT"S A MAGIC NUMBER THAT MEANS RIDICULOUS, GIMME A BREAK

Reply:
we bail out of this loop when everything works, so we never actually hit 10000 unless the test is failing

I see your valid concern